src/librustc/ty/error.rs

```diff
@@ -273,8 +275,83 @@ impl<'tcx> TyCtxt<'tcx> {
                                 Applicability::MachineApplicable
                             );
                         }
+                    },
+                    (ty::Param(_), ty::Param(_)) => {
+                        db.note("a type parameter was expected, but a different one was found; \
+                                 you might be missing a type parameter or trait bound");
+                        db.note("for more information, visit \
+                                 https://doc.rust-lang.org/book/ch10-02-traits.html\
+                                 #traits-as-parameters");
+                    }
+                    (ty::Projection(_), ty::Projection(_)) => {
+                        db.note("an associated type was expected, but a different one was found");
+                    }
+                    (ty::Param(_), ty::Projection(_)) | (ty::Projection(_), ty::Param(_)) => {
+                        db.note("you might be missing a type parameter or trait bound");
+                    }
+                    (ty::Param(_), _) | (_, ty::Param(_)) => {
+                        db.help("given a type parameter `T` and a method `foo`:
```

Comment:
Can we gate the long explanations (both of them) under the teaching flag?

Reply:
I just think we need to keep in mind folks who have already learned the concept and don't want to be spammed down with very long messages (which are useful to beginners but not familiar folks).

Reply:
Agree.

I think it would be better if we just had a section in the book that we could link to explaining this case. Having at most two lines, including link, of text should be reasonable (and is my unofficial cutoff for these kind of messages).

What do you think?

Reply:
Sgtm